File: Problems/Parking Lot System/ParkingManager.cpp
#pragma once
#include "./Vechicle.cpp"
#include "./Ticket.cpp"
#include "./Buldings.cpp"
#include "./Floors.cpp"
#include  "./ParkingSpots.cpp"

class ParkingManager{
    vector<Building*>buildings; // Changed to store pointers
    unordered_map<int,Ticket*> tickets;
    int ticketCounter = 0;
    public:

    ParkingManager() = default;

    void addBuilding(Building* building) { // Changed to accept pointer
        buildings.push_back(building);
    }
    void removeBuilding(int buildingId) {
        buildings.erase(remove_if(buildings.begin(), buildings.end(),
            [buildingId](Building* b) { // Changed to work with pointer
                return b->getBuildingId() == buildingId; // Use -> instead of .
            }), buildings.end());
    }

    ParkingSpot* findAvailableSpot(VehicleType vehicleType) {
        for (auto& building : buildings) {
            ParkingSpot* spot = building->findParkingSpot(vehicleType); // Pass vehicle type
            if (spot != nullptr) {
                return spot; // Return the first available parking spot found
            }
        }
        return nullptr; // No available parking spots in any building
    }

    Ticket* parkVehicle(Vehicle* vehicle) {
        ParkingSpot* spot = findAvailableSpot(vehicle->getType()); // Pass vehicle type
        if (spot == nullptr) {
            throw runtime_error("No available parking spots for this vehicle type.");
        }

        spot->assignVehicle(vehicle);
        Ticket* ticket = new Ticket(++ticketCounter, vehicle, spot);
        tickets[ticketCounter] = ticket;
        return ticket;
    }

    void removeVehicle(int ticketId) {
        auto it = tickets.find(ticketId);
        if (it == tickets.end()) {
            throw runtime_error("Ticket not found.");
        }

        Ticket* ticket = it->second;
        ParkingSpot* spot = ticket->getSpot();
        spot->removeVehicle();
        delete ticket; // Free the ticket memory
        tickets.erase(it); // Remove the ticket from the map
    }
};

File: Problems/Parking Lot System/ParkingManager.java
package Problems.Java;

public class ParkingManager {
    ParkingLot parkingLot;
    TicketManager ticketManager;


    public ParkingManager(ParkingLot parkingLot) {
        this.parkingLot = parkingLot;
    }

    Ticket parkVehicle(Vehicle vehicle) {
        Spot parkingSpot = parkingLot.parkVehicle(vehicle);
        if(parkingSpot == null) {
            System.out.println("Sorry, parking lot is full");
            return null;
        }
        Ticket ticket = ticketManager.getTicket(vehicle,parkingSpot, 0);
        // we will only pass the vehicle
        // it will find the suitable parking spot
        // it will park the vehicle and generate a ticket and return it.
        return ticket;
    }
    public double unParkVehicle(Ticket ticket) {
        // it will go through this parkingticket and unpark the vehicle
        boolean isUnParked = parkingLot.unParkVehicle(ticket.getSpot());
        if(isUnParked) {
            return ticketManager.calculateFare(ticket,0);
        }

        return 0;
    }

    
    
}
